### PR TITLE
parallelize repository mirroring

### PR DESCRIPTION
**Pull request by vladak**
will be handy for https://github.com/OpenGrok/docker/pull/32

### CODE REVIEW

File: opengrok-tools/src/main/python/opengrok_tools/mirror.py [hunk 1]
```diff
@@ -52,7 +53,37 @@
...
+def worker(cfg):
+    if cfg.batch:
+        get_batch_logger(cfg.logdir, cfg.project_name,
+                         cfg.loglevel,
+                         cfg.backupcount,
+                         get_class_basename())
+
+    return mirror_project(cfg.config, cfg.project_name,
+                          cfg.incoming,
+                          cfg.uri, cfg.source_root)
```
Comment (tulinkry) [inline opengrok-tools/src/main/python/opengrok_tools/mirror.py]: I personally developed a decorator for pool map function, you can use it if you feel it's better

```
def pool_callback(callback):
    """
    Decorator for auto expanding the thread pool arguments from a single
    variable list to a function arguments.
    :param callback: the actual executing method
    :return: a function calling the callback with proper argument list
    """
    return lambda args: callback(*args)
```

Usage
```
@pool_callback
def worker(project, logdir, loglevel, ...):
    pass
```

File: opengrok-tools/src/main/python/opengrok_tools/mirror.py [hunk 2]
```diff
@@ -34,6 +34,7 @@
 import os
 import sys
 import tempfile
+from multiprocessing import Pool, cpu_count
```
Comment (tulinkry) [inline opengrok-tools/src/main/python/opengrok_tools/mirror.py]: is process pool better than thread pool?

Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/mirror.py]: I think in this case it does not matter as calling `sync()` from `scm` modules will result in process spawning.

File: opengrok-tools/src/main/python/opengrok_tools/mirror.py (hunk 1 above)
Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/mirror.py]: So the 2nd arg to `pool.map` is then list of lists ?

Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/mirror.py]: Yes

Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/mirror.py]:
```
@pool_callback
def callback(arg1, arg2, arg3):
    return arg1, arg2, arg3

ThreadPool(1).map(callback, [[1, 2, 3]])
```

Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/mirror.py]: This does not work with regular `Pool`:
```python
with Pool(processes=int(1)) as pool:                                            
    ret = pool.map(callback, [[1, 2, 3], [4, 5, 6]], 1)
```
which fails with:
```
Traceback (most recent call last):
  File "./pool_callb.py", line 25, in <module>
    ret = pool.map(callback, [[1, 2, 3], [4, 5, 6]], 1)
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 288, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 670, in get
    raise self._value
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 450, in _handle_tasks
    put(task)
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
AttributeError: Can't pickle local object 'pool_callback.<locals>.<lambda>'
```
I think this is due to the passing/serialization of worker arguments to worker processes. Looks like it wants to send the lambda across and use it only when it reaches the worker.

Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/mirror.py]: In the end I unpacked the list manually.

Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/mirror.py]: ah so, the interprocess communication is little bit different, ok.

**Pull request merged**